cmd/nvidia-device-plugin/prestart-hook.go

```diff
@@ -0,0 +1,283 @@
+
+	cmd := hook.preStartCmd[0]
+	args := hook.preStartCmd[1:]
+	for _, device := range devices {
+		uuid := strings.TrimPrefix(device.ID, "GPU-")
+		args = append(args, uuid)
+	}
+
+	log.Printf("PreStartHook(%s): creating command %s %v\n", logCtx, cmd, args)
+	cmdCtx := exec.CommandContext(ctx, cmd, args...)
+	log.Printf("PreStartHook(%s): created command\n", logCtx)
+
+	stdout, err := cmdCtx.StdoutPipe()
+	if err != nil {
+		err = fmt.Errorf("error getting stdout pipe: %v", err)
+		return
+	}
+	stdoutDone := make(chan struct{})
+	go func() {
+		s := bufio.NewScanner(stdout)
+		for s.Scan() {
+			line := s.Text()
+			fmt.Printf("PreStartHook(%s)::out: %s\n", logCtx, line)
+		}
+		stdoutDone <- struct{}{}
+	}()
+	log.Printf("PreStartHook(%s): got stdout\n", logCtx)
+
+	stderr, err := cmdCtx.StderrPipe()
+	if err != nil {
+		err = fmt.Errorf("error getting stderr pipe: %v", err)
+		return
+	}
+	stderrDone := make(chan []string)
+	go func() {
+		s := bufio.NewScanner(stderr)
+		uuids := []string{}
+		for s.Scan() {
+			line := s.Text()
+			log.Printf("PreStartHook(%s)::err: %s\n", logCtx, line)
+			uuids = append(uuids, line)
+		}
+		stderrDone <- uuids
+	}()
+	log.Printf("PreStartHook(%s): got stderr\n", logCtx)
+
+	err = cmdCtx.Start()
+	if err != nil {
+		err = fmt.Errorf("error starting: %v", err)
+		return
+	}
+	log.Printf("PreStartHook(%s): started command\n", logCtx)
+
+	err = cmdCtx.Wait()
+	if err != nil {
+		err = fmt.Errorf("command failed to run: %v", err)
+	}
+	log.Printf("PreStartHook(%s): command exited with code %d\n", logCtx, cmdCtx.ProcessState.ExitCode())
+
+	<-stdoutDone
+	log.Printf("PreStartHook(%s): stdout closed\n", logCtx)
+
+	unhealthy = <-stderrDone
+	log.Printf("PreStartHook(%s): stderr closed\n", logCtx)
+
+	return
+}
+
+func markUnhealthy(device *rm.Device, unhealthyChan chan<- *rm.Device, logCtx string) {
+	log.Printf("PreStartHook(%s): Marking device %s as unhealthy\n", logCtx, deviceString(device))
+	unhealthyChan <- device
+}
+
+func getLogContextFromDeviceIDs(deviceIDs []string) string {
+	return strings.ReplaceAll(strings.Join(deviceIDs, "-"), "nvidia", "")
+}
+
+func getLogContextFromDevices(devices []*rm.Device) string {
+	ids := make([]string, len(devices))
+	for i, v := range devices {
+		ids[i] = v.ID
+	}
+	return getLogContextFromDeviceIDs(ids)
+}
+
+func (hook *PreStartHook) getDevicePreStartInfo(ctx context.Context, deviceId string, logCtx string) (*DevicePreStartInfo, error) {
+	err := hook.deviceInfosLock.Acquire(ctx, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	defer hook.deviceInfosLock.Release(1)
+	info, ok := hook.deviceInfos[deviceId]
+	if !ok {
+		log.Printf("PreStartHook(%s): Creating device PreStart info for %q\n", logCtx, deviceId)
+		info = newDevicePreStartInfo()
+		hook.deviceInfos[deviceId] = info
+	}
+	return info, nil
+}
+
+func findDevice(deviceId string, allDevices []*rm.Device) (*rm.Device, error) {
+	for _, device := range allDevices {
+		if device.ID == deviceId {
```

Comment:
Not that with GPU sharing enabled, `device.ID` may not be the `uuid` of the device. There is ` GetUUID` method that can be used instead.